From 3434a21151055b597915f6ff94255a1a195a9ed5 Mon Sep 17 00:00:00 2001
From: Joshua Haberman <haberman@google.com>
Date: Wed, 16 Apr 2025 13:03:08 -0700
Subject: [PATCH] Fixed LTO-only linker error in upb linker arrays.

Also removed a sanitizer warning on non-Clang compilers.

Fixes: https://github.com/protocolbuffers/protobuf/issues/21021
PiperOrigin-RevId: 748395156
---
 upb/mini_table/extension_registry.c |  2 +-
 upb/port/def.inc                    | 29 ++++++++++++++++++-----------
 2 files changed, 19 insertions(+), 12 deletions(-)

diff --git a/upb/mini_table/extension_registry.c b/upb/mini_table/extension_registry.c
index 65ac787..ff33ec5 100644
--- a/upb/mini_table/extension_registry.c
+++ b/upb/mini_table/extension_registry.c
@@ -72,7 +72,7 @@ failure:
 
 #ifdef UPB_LINKARR_DECLARE
 
-UPB_LINKARR_DECLARE(upb_AllExts, upb_MiniTableExtension);
+UPB_LINKARR_DECLARE(upb_AllExts, const upb_MiniTableExtension);
 
 bool upb_ExtensionRegistry_AddAllLinkedExtensions(upb_ExtensionRegistry* r) {
   const upb_MiniTableExtension* start = UPB_LINKARR_START(upb_AllExts);
diff --git a/upb/port/def.inc b/upb/port/def.inc
index 4c073b3..e09e9ec 100644
--- a/upb/port/def.inc
+++ b/upb/port/def.inc
@@ -354,6 +354,12 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
 
 #undef UPB_IS_GOOGLE3
 
+#ifdef __clang__
+#define UPB_NO_SANITIZE_ADDRESS __attribute__((no_sanitize("address")))
+#else
+#define UPB_NO_SANITIZE_ADDRESS
+#endif
+
 // Linker arrays combine elements from multiple translation units into a single
 // array that can be iterated over at runtime.
 //
@@ -381,10 +387,10 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
 #if defined(__ELF__) || defined(__wasm__)
 
 #define UPB_LINKARR_APPEND(name) \
-  __attribute__((retain, used, section("linkarr_" #name)))
+  __attribute__((retain, used, section("linkarr_" #name))) UPB_NO_SANITIZE_ADDRESS
 #define UPB_LINKARR_DECLARE(name, type)     \
-  extern type const __start_linkarr_##name; \
-  extern type const __stop_linkarr_##name;  \
+  extern type __start_linkarr_##name; \
+  extern type __stop_linkarr_##name;  \
   UPB_LINKARR_APPEND(name) type UPB_linkarr_internal_empty_##name[1]
 #define UPB_LINKARR_START(name) (&__start_linkarr_##name)
 #define UPB_LINKARR_STOP(name) (&__stop_linkarr_##name)
@@ -393,13 +399,13 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
 
 /* As described in: https://stackoverflow.com/a/22366882 */
 #define UPB_LINKARR_APPEND(name) \
-  __attribute__((retain, used, section("__DATA,__la_" #name)))
-#define UPB_LINKARR_DECLARE(name, type)           \
-  extern type const __start_linkarr_##name __asm( \
-      "section$start$__DATA$__la_" #name);        \
-  extern type const __stop_linkarr_##name __asm(  \
-      "section$end$__DATA$"                       \
-      "__la_" #name);                             \
+  __attribute__((retain, used, section("__DATA,__la_" #name))) UPB_NO_SANITIZE_ADDRESS
+#define UPB_LINKARR_DECLARE(name, type)     \
+  extern type __start_linkarr_##name __asm( \
+      "section$start$__DATA$__la_" #name);  \
+  extern type __stop_linkarr_##name __asm(  \
+      "section$end$__DATA$"                 \
+      "__la_" #name);                       \
   UPB_LINKARR_APPEND(name) type UPB_linkarr_internal_empty_##name[1]
 #define UPB_LINKARR_START(name) (&__start_linkarr_##name)
 #define UPB_LINKARR_STOP(name) (&__stop_linkarr_##name)
@@ -414,7 +420,8 @@ void __asan_unpoison_memory_region(void const volatile *addr, size_t size);
 // Usage of __attribute__ here probably means this is Clang-specific, and would
 // not work on MSVC.
 #define UPB_LINKARR_APPEND(name) \
-  __declspec(allocate("la_" #name "$j")) __attribute__((retain, used))
+  __declspec(allocate("la_" #name "$j")) \
+  __attribute__((retain, used)) UPB_NO_SANITIZE_ADDRESS
 #define UPB_LINKARR_DECLARE(name, type)                               \
   __declspec(allocate("la_" #name "$a")) type __start_linkarr_##name; \
   __declspec(allocate("la_" #name "$z")) type __stop_linkarr_##name;  \
-- 
2.51.1

